File: prefix suffix sum/prefix_suffix_algorithms/practice/range_sum_query_immutable.py
"""
Problem Link: https://leetcode.com/problems/range-sum-query-immutable/
Difficulty: Easy

Problem Description:
Design a data structure that handles multiple queries of the following type:
1. `sumRange(left, right)`: Calculate the sum of the elements of `nums` between indices `left` and `right` inclusive.

Key Concepts:
- Prefix Sum Array
- Preprocessing vs Query Time Trade-off

Intuition:
If we just iterate from `left` to `right` for every query, the time taken is proportional to the length of the range (O(N)).
If we have many queries (say M queries), the total time would be O(M * N), which is too slow.

By pre-calculating the running totals (prefix sums) once, we can answer any range sum query in O(1) time.
`Sum(left, right) = PrefixSum[right] - PrefixSum[left - 1]`

Algorithm:
1. **Initialization (__init__)**:
   - Create a `prefix` array of size `N + 1`.
   - We use size `N + 1` to handle the edge case where `left = 0` elegantly (so we don't index out of bounds).
   - `prefix[i]` will store the sum of `nums[0]` up to `nums[i-1]`.
   - `prefix[0]` will be 0.

2. **Query (sumRange)**:
   - Return `prefix[right + 1] - prefix[left]`.

Time Complexity:
- Initialization: O(N) to build the array.
- Query: O(1) constant time per lookup.

Space Complexity: O(N) to store the prefix sum array.
"""

from typing import List

class NumArray:

    def __init__(self, nums: List[int]):
        # Initialize prefix array with N+1 zeros
        # self.prefix[k] covers sum of nums[0...k-1]
        self.prefix = [0] * (len(nums) + 1)
        
        # Build the prefix sum array
        for i in range(len(nums)):
            # New prefix sum = Previous prefix sum + Current number
            self.prefix[i + 1] = self.prefix[i] + nums[i]

    def sumRange(self, left: int, right: int) -> int:
        # Formula: P[right+1] - P[left]
        # P[right+1] includes nums[0]...nums[right]
        # P[left] includes nums[0]...nums[left-1]
        # Subtracting gives nums[left]...nums[right]
        return self.prefix[right + 1] - self.prefix[left]


# ---------------------------------------------------
# Test Cases
# ---------------------------------------------------
if __name__ == "__main__":
    nums = [-2, 0, 3, -5, 2, -1]
    obj = NumArray(nums)
    
    print(f"Array: {nums}")
    print(f"Sum(0, 2): {obj.sumRange(0, 2)} (Expected: 1)") # -2 + 0 + 3 = 1
    print(f"Sum(2, 5): {obj.sumRange(2, 5)} (Expected: -1)") # 3 + -5 + 2 + -1 = -1
    print(f"Sum(0, 5): {obj.sumRange(0, 5)} (Expected: -3)") 

File: prefix suffix sum/prefix_suffix_algorithms/practice/range_sum_query_immutable.js
/**
 * Problem Link: https://leetcode.com/problems/range-sum-query-immutable/
 * Difficulty: Easy
 * 
 * Problem Description:
 * Design a class that handles multiple range sum queries efficiently.
 * `sumRange(left, right)` should return the sum of elements between `left` and `right`.
 * 
 * Key Concepts:
 * - Prefix Sum Array
 * - Class / Object-Oriented Design
 * 
 * Intuition:
 * Naive loop summation is O(N) per query. We want O(1).
 * We pre-calculate cumulative sums so that any range sum is just a subtraction of two pre-calculated values.
 * 
 * Algorithm:
 * 1. Constructor: Build an array `prefix` where `prefix[i]` stores the sum of the first `i` elements.
 * 2. `prefix[0]` is always 0 (identity for addition).
 * 3. `sumRange(L, R)` is simply `prefix[R + 1] - prefix[L]`.
 * 
 * Time Complexity: O(N) init, O(1) query.
 * Space Complexity: O(N) storage.
 */

class NumArray {
    /**
     * @param {number[]} nums
     */
    constructor(nums) {
        // Create an array of size N+1 filled with 0
        this.prefix = new Array(nums.length + 1).fill(0);

        // Fill the prefix array
        for (let i = 0; i < nums.length; i++) {
            // current prefix sum = previous sum + current number
            this.prefix[i + 1] = this.prefix[i] + nums[i];
        }
    }

    /**
     * @param {number} left 
     * @param {number} right 
     * @return {number}
     */
    sumRange(left, right) {
        // To get sum inclusive of right, we need index right+1
        // To exclude everything before left, subtract index left
        return this.prefix[right + 1] - this.prefix[left];
    }
}

// ---------------------------------------------------
// Test Cases
// ---------------------------------------------------
const nums = [-2, 0, 3, -5, 2, -1];
const obj = new NumArray(nums);

console.log(`Array: [${nums}]`);
console.log(`Sum(0, 2): ${obj.sumRange(0, 2)} (Expected: 1)`);
console.log(`Sum(2, 5): ${obj.sumRange(2, 5)} (Expected: -1)`);
console.log(`Sum(0, 5): ${obj.sumRange(0, 5)} (Expected: -3)`);
